pest/src/error.rs

```diff
@@ -317,6 +404,100 @@ impl<R: RuleType> Error<R> {
         self
     }
 
+    /// Get detailed information about errored rules sequence.
+    /// Returns `Some(results)` only for `ParsingError`.
+    pub fn parse_attempts(&self) -> Option<ParseAttempts<R>> {
+        self.parse_attempts.clone()
+    }
+
+    /// Get error message based on parsing attempts.
+    /// Returns `None` in case self `parse_attempts` is `None`.
+    pub fn parse_attempts_error(
+        &self,
+        input: &str,
+        rule_to_message: &RuleToMessageFn<R>,
+        is_whitespace: &IsWhitespaceFn,
+    ) -> Option<Error<R>> {
+        let attempts = if let Some(ref parse_attempts) = self.parse_attempts {
+            parse_attempts.clone()
+        } else {
+            return None;
+        };
+
+        let spacing = self.spacing() + "   ";
+        let error_position = attempts.max_position;
+        let message = {
+            let mut help_lines: Vec<String> = Vec::new();
+            help_lines.push(String::from("error: parsing error occurred."));
+
+            // Note: at least one of `(un)expected_tokens` must not be empty.
+            if let Some(m) = attempts.tokens_message(is_whitespace, true, &spacing) {
+                help_lines.push(m)
+            };
+            if let Some(m) = attempts.tokens_message(is_whitespace, false, &spacing) {
+                help_lines.push(m)
+            };
```

Comment:
Given that this is the only place `tokens_message` is called, is there any benefit to splitting by "expected"? Maybe just leave out the `expected` boolean arg and return both.